slurm_ops_manager/slurm_ops_base.py

```diff
@@ -267,13 +270,104 @@ def upgrade(self):
         """Preform upgrade-charm operations."""
         raise Exception("Inheriting object needs to define this method.")
 
-    def setup_system(self):
-        """Preform the install and setup operations."""
+    def setup_slurm(self):
+        """Install and setup Slurm and its dependencies."""
         raise Exception("Inheriting object needs to define this method.")
 
+    @property
+    def nhc_version(self) -> str:
+        """Return NHC version."""
+        return "1.4.2-omni-1.0"
+
+    def _install_nhc_from_git(self) -> None:
+        """Install NHC from Omnivector fork."""
+        version = self.nhc_version
+        src = f"https://codeload.github.com/omnivector-solutions/nhc/tar.gz/refs/tags/{version}"
+
+        logger.info(f"#### downloading and installing NHC {version}")
+
+        base_path = Path("/tmp/nhc")
+        full_path = base_path / f"nhc-{version}"
+        nhc_tar = base_path / "nhc.tar.gz"
+
+        # cleanup old installations
+        if base_path.exists():
+            rmtree(base_path)
+        base_path.mkdir()
+
+        cmd = f"curl -o {nhc_tar} -s {src}".split()
+        subprocess.run(cmd)
+
+        cmd = f"tar --extract --directory {base_path} --file {nhc_tar}".split()
+        subprocess.run(cmd)
+
+        if operating_system() == 'ubuntu':
+            libdir = "/usr/lib"
+        else:
+            libdir = "/usr/libexec"
+
+        # NOTE: this requires make. We install it using the dispatch file in
+        # the slurmd charm.
+        try:
+            locale = {'LC_ALL':'C', 'LANG':'C.UTF-8'}
+            cmd = f"./autogen.sh --prefix=/usr --sysconfdir=/etc \
+                                 --libexecdir={libdir}".split()
+            logger.info(f'##### NHC - running autogen')
+            r = subprocess.run(cmd, cwd=full_path, env=locale,
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.STDOUT)
+            logger.debug(f'##### autogen: {r.stdout.decode()}')
+            r.check_returncode()
+
+            logger.info(f'##### NHC - running tests')
+            r = subprocess.run(["make", "test"], cwd=full_path,
+                               env=locale, stdout=subprocess.PIPE,
+                               stderr=subprocess.STDOUT)
+            logger.debug(f'##### NHC make test: {r.stdout.decode()}')
+            r.check_returncode()
+            assert "tests passed" in r.stdout.decode()
```

Comment:
We don't catch this assertion in the case it doesn't pass  .... possibly catch the exception if it doesn't pass or return -1. I would prefer not to have assertions in the code, outside of tests.

Reply:
@jamesbeedy updated the codewith your suggestion